docs/latest/design/extending-envoy-gateway.md

```diff
@@ -0,0 +1,176 @@
+
+An example configuration:
+
+```yaml
+apiVersion: config.gateway.envoyproxy.io/v1alpha1
+kind: EnvoyGateway
+extensions:
+  - name: example-extension
+    apiGroups:
+      - apiGroup: example.myextension.io
+        kinds:
+          - OAuth2Filter
+    service:
+      host: my-extension.example
+      port: 443
+      tls:
+        type: secret
+        secret:
+            name: my-extension
+            namespace: example
+```
+
+This configuration is required to be provided at bootstrap and modifying the registered extensions during runtime is not currently supported.
+Envoy Gateway will keep track of the registered extensions and their `apiGroups` and `kinds` when processing Gateway API resources.
+
+## Extending Gateway API and the Data Plane
+
+Envoy Gateway manages [Envoy][] deployments, which act as the data plane that handles actual user traffic. Users configure the data plane using the K8s Gateway API resources which Envoy 
+Gateway converts into [Envoy specific configuration (xDS)][] to send over to Envoy.
+
+Gateway API offers [ExtensionRef filters][] and [Policy Attachments][] as extension points for implementers to use. Envoy Gateway extends the Gateway API using these extension points to provide support for [rate limiting][] 
+and [authentication][]. The initial design of Envoy Gateway extensions will primarily focus on ExtensionRef filters, where users can reference a custom resource as an HTTP Filter.
+
+When Envoy Gateway encounters an [HTTPRoute][] or [GRPCRoute][] that has an `ExtensionRef` `filter` with an `apiGroup` and `kind` that Envoy Gateway does not supoport, it will first
+check the list of registered extensions to determine if any of the registered extensions support the referenced object before considering it a configuration error.
+
+This allows users to be able to reference additional filters provided by extensions to Envoy Gateway, in their `HTTPRoute`s / `GRPCRoute`s:
+
+```yaml
+apiVersion: example.myextension.io/v1alpha1
+kind: OAuth2Filter
+metadata:
+  name: oauth2-filter
+spec:
+  ...
+
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: example
+spec:
+  parentRefs:
+  - name: eg
+  hostnames:
+  - www.example.com
+  rules:
+  - clientSelectors:
+    - path:
+        type: PathPrefix
+        value: /
+    filters:
+    - type: ExtensionRef
+      extensionRef:
+        group: example.myextension.io
+        kind: OAuth2Filter
+        name: oauth2-filter
+    backendRefs:
+    - name: backend
+      port: 3000
+```
+
+In order to enable the translation of new resources introduced by extensions, Envoy Gateway provides hook points within the translation pipeline, where it calls out to extension services registered in the [EnvoyGateway config][] 
+if they specify an `apiGroup` that matches the `group` of an `ExtensionRef` filter. The extension will then be able to modify the xDS that Envoy Gateway generated and send back the 
```

Comment:
Does it mean that the extension will work in the xDS Translator layer?
Envoy gateway will send the xDS IR to the extension and the extension send the xDS patch back to the Envoy gateway.

Reply:
Correct. For the initial implementation of this design, there will likely only be a single hook point where an extension has control over the xDS resources right before Envoy Gateway updates them in the xDS server to be consumed by Envoy Proxy. Still, the framework of the extensions system is meant to be a foundation where more fine-grained hooks can be introduced into the future for use cases where the extension does not need such broad control.

Reply:
Thanks for your reply.